### DISCUSSION TITLE
Plain WPF sample doesn't work - Cef::Initialize() failed

### DISCUSSION THREAD

**Discussion by WrongBit** [cefsharp/CefSharp#3417] [General]
Hi! I have Win7x64, FW4.8, VS2019
I tried to create simple example(AnyCPU + **not** "prefer 32 bits"), but w/o any clownish nugets - we have strict policy to not deal with untested updates.
I unpacked all nuget packages and got x64 DLLs. Then I referenced DLLs in my WPF project + added necessary native libs (which come with cef.redist.x64.88.2.4.nupkg ). XAML code is quite simple:

```
<Window x:Class="ChromeEmbFrw.MainWindow" xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation" 
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:cef="clr-namespace:CefSharp.Wpf;assembly=CefSharp.Wpf"
        Title="MainWindow" Height="450" Width="800">
    <cef:ChromiumWebBrowser Name="brwMain" />
</Window>

```

And when I start application, got this error:

    Cef::Initialize() failed
    at CefSharp.Wpf.ChromiumWebBrowser.NoInliningConstructor() in 
    C:\\projects\\cefsharp\\CefSharp.Wpf\\ChromiumWebBrowser.cs:line 530
    at CefSharp.Wpf.ChromiumWebBrowser..ctor() in C:\\projects\\cefsharp\\CefSharp.Wpf\\ChromiumWebBrowser.cs:line 451

For me it's quite strange to see so unprofessional, obscure reports like "failed" - I have no idea what exactly going wrong.
Of course I tried to inspect `ChromiumWebBrowser.cs:line 530`, but found nothing, but empty line:

```
IsVisibleChanged += OnIsVisibleChanged;
// line 530
ToolTip = toolTip = new ToolTip();

```
Hardly assigning handler was reason, but anyway I cannot debug it.
This is list of binaries I got after build (and added from redist):

```
CefSharp.BrowserSubprocess.Core.dll
CefSharp.BrowserSubprocess.Core.pdb
CefSharp.BrowserSubprocess.exe
CefSharp.BrowserSubprocess.pdb
CefSharp.Core.dll
CefSharp.Core.pdb
CefSharp.Core.Runtime.dll
CefSharp.Core.Runtime.pdb
CefSharp.Core.Runtime.xml
CefSharp.Core.xml
CefSharp.dll
CefSharp.pdb
CefSharp.Wpf.dll
CefSharp.Wpf.pdb
CefSharp.Wpf.xml
CefSharp.xml
ChromeEmbFrw.exe   - my app
ChromeEmbFrw.exe.config
ChromeEmbFrw.pdb
chrome_elf.dll
icudtl.dat
libcef.dll
snapshot_blob.bin
v8_context_snapshot.bin

```

I debugged everything I could (I had no necessary files), but now it's complete set of binaries and I have no idea WHAT ELSE can be wrong. So I need help please what to do. Ideally if people takes "spaghetti", which produces error and UNHIDE internal problems - no any reason to hide it even from final user! We always must know what's wrong.
Thanks!

**Comment by amaitland** [comment]
CEF doesn't provide any error information,  we only know that the call to [CefInitialize](https://magpcss.org/ceforum/apidocs3/projects/(default)/(_globals).html#CefInitialize(constCefMainArgs&,constCefSettings&,CefRefPtr%3CCefApp%3E,void*)) failed. This is a limitation of the Chromium embedded framework(CEF).

You'll need to check the CEF log file see https://github.com/cefsharp/CefSharp/wiki/Trouble-Shooting#log-file

You are likely missing files from your bin folder, see https://github.com/chromiumembedded/cef/blob/cae194f493328a9890b27533ce1dc661e0d4b042/tools/distrib/win/README.redistrib.txt

Whilst CEF lists components as optional I expect you'll need to include most of the pak files. Once you have an error from the log you can search on https://magpcss.org/ceforum/index.php (The CEF support forum) to see what's required, I personally suggest including all of the resources excluding any locales that aren't relevant. 

Using CefSharp via the Nuget packages is the only supported option, if you choose to manually package CEF then you'll need to deal with the complexities yourself.

**Reply by amaitland** [reply]
For background please read the first three paragraphs of https://github.com/cefsharp/CefSharp/wiki/General-Usage

**Comment by WrongBit** [comment]
@amaitland : thank you for important info! I missed "init" requirement:

        public App()
        {
            var settings = new CefSettings() { CachePath = @"R:\6", LogFile = @"c:\DEV\browser_err.log", LogSeverity = LogSeverity.Verbose };
            //Perform dependency check to make sure all relevant resources are in our output directory.
            Cef.Initialize(settings, performDependencyCheck: true, browserProcessHandler: null);
        }

But even that didn't help even a bit. Why? Log file wasn't created! I run app, SAME exception happen and... nothing - no any log.
Also I used your advice to put all redistributable files - also exception.
I'm done - I cannot fight with ugly designed rubbish. Anyway thanks for help!

**Reply by amaitland** [reply]
There's a reason the project is packaged via Nuget,  including CEF in a .Net app is fiddly.

https://github.com/cefsharp/CefSharp.MinimalExample provides a working example.

**Reply by WrongBit** [reply]
There is no reason for nuget AT ALL. It's ugly, dangerous technology, coming from "linux packages". Enterprise world cannot work on this parody.
